File: BackEnd/classes/stag.py
import requests
import os
from lib.HTTP_messages import internal_server_error
from logging import ERROR, log


def get(ticket, url, params):
    response = None
    try:
        headers = {
            "accept": "application/json",
            "Content-Type": "application/json",
            "Connection": "keep-alive",
            "Accept-Origin": os.getenv("STAG_URL"),
        }
        url = os.getenv('STAG_URL') + url
        response = requests.get(url, params=params, headers=headers, cookies={'WSCOOKIE': ticket})
        if not response.ok:
            raise Exception(response.text)
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    if not response:
        raise Exception("missing response from stag")
    try:
        response = response.json()
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    return response


def get_stag_user_info(ticket):
    """ Vrátí jméno, příjmení, email, titul a stagUserInfo (username, role, nazev, ucitIdno/osCilo, email)"""
    response = None
    try:
        envUrl = os.getenv('STAG_URL')
        if not envUrl:
            raise Exception("missing env val")
        url = envUrl + "ws/services/rest2/help/getStagUserListForLoginTicketV2?ticket=" + ticket
        headers = {
            "accept": "application/json",
            "Content-Type": "application/json",
            "Connection": "keep-alive",
            "Accept-Origin": os.getenv("STAG_URL"),
        }
        response = requests.get(url, headers=headers, cookies={'WSCOOKIE': ticket})
        if not response.ok:
            return internal_server_error
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    if not response:
        raise Exception("missing response from stag")
    try:
        response = response.json()
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    return response


def bool_existuje_predmet(ticket, katedra, zkratka_predmetu):
    """ Vrátí informace o předmětu """
    response = None
    try:
        envUrl = os.getenv('STAG_URL')
        if not envUrl:
            raise Exception("missing env val")
        url = envUrl + "ws/services/rest2/predmety/getPredmetInfo"
        headers = {
            "accept": "application/json",
            "Content-Type": "application/json",
            "Connection": "keep-alive",
            "Accept-Origin": os.getenv("STAG_URL"),
        }
        params = {
            "katedra": katedra,
            "zkratka": zkratka_predmetu
        }
        response = requests.get(url, headers=headers, params=params)
        if not response.ok:
            return None
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    if not response:
        raise Exception("missing response from stag")
    try:
        response = response.json()
        if response:
            return True
        else:
            return False
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error


def get_vyucujici_predmetu_stag(zkratka_predmetu, katedra):
    """ Vrátí informace o predmetu"""
    response = None
    try:
        envUrl = os.getenv('STAG_URL')
        if not envUrl:
            raise Exception("missing env val")
        params = {
            "katedra": katedra,
            "zkratka": zkratka_predmetu
        }
        url = envUrl + "ws/services/rest2/predmety/getPredmetInfo"
        headers = {
            "accept": "application/json",
            "Content-Type": "application/json",
            "Connection": "keep-alive",
            "Accept-Origin": os.getenv("STAG_URL"),
        }
        response = requests.get(url, headers=headers, params=params)
        if not response.ok:
            return "chyba"
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error
    if not response:
        raise Exception("missing response from stag")
    try:
        response = response.json()
        return response["cvicici"]
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error


def get_userid_and_role(json):
    """Vrací userId a roli uživatele
    json: json, který vrací funkce "get_stag_user_info"
    """
    try:
        role = json["stagUserInfo"][0]["role"]
        if role == "":
            return internal_server_error, internal_server_error
        if "ST" not in role:
            userid = str(json["stagUserInfo"][0]["ucitIdno"])
        else:
            userid = str(json["stagUserInfo"][0]["osCislo"])
        return userid, role
    except KeyboardInterrupt:
        os.close(1)
    except SystemExit:
        os.close(1)
    except Exception as e:
        log(ERROR, e)
        return internal_server_error, internal_server_error


__all__ = ["get", "get_stag_user_info", "bool_existuje_predmet", "get_vyucujici_predmetu_stag", "get_userid_and_role"]

File: src/lib/stag.ts
import {
 tGetPredmetInfo,
 tGetStudentInfo,
 tGetStagUserListForLoginTicketV2,
 tGetStudentiByPredmet,
 tGetRozvrhByStudent,
} from '@/types/stag_response_types'
import crypto from 'crypto'

/* Header Based ticket in X-Stag-Ticket */
export function getTicketX(req: Request): string | null {
 const headers = req.headers
 const rTicket = headers.get('x-svt') || ''

 if (!rTicket) {
  return null
 }
 return rTicket
}

// Header Based ticket in Cookie
export function encodeId(id: string) {
 const hash = crypto.createHash('sha1')
 hash.update(id)
 return hash.digest('hex')
}

function assembleHeaders(ticket: string): Headers {
 const headers = new Headers({
  accept: 'application/json',
  'Content-Type': 'application/json',
  Connection: 'keep-alive',
  'Accept-Origin': `${process.env.STAG_SERVER}`,
 })
 headers.set('Cookie', `WSCOOKIE=${ticket}`)
 return headers
}

/*
 * STAG API calls
 *
 * All functions return null if anything goes wrong (non-200 response, empty response, etc.)
 */

export async function getUserInfo(
 ticket: string,
): Promise<tGetStagUserListForLoginTicketV2 | null> {
 const headers = assembleHeaders(ticket)

 const res = await fetch(
  `${process.env.STAG_SERVER}/services/rest2/help/getStagUserListForLoginTicketV2?ticket=${ticket}`,
  { method: 'GET', headers },
 )
 if (!res.ok) return null
 if (res.status === 204) return null

 const data = (await res.json()) as tGetStagUserListForLoginTicketV2
 if (!data) return null

 for (const user of data.stagUserInfo) {
  if (user.osCislo) user.encId = encodeId(`${user.osCislo}`)
  if (user.ucitIdno) user.encId = encodeId(`${user.ucitIdno}`)
 }

 return data
}

export async function getStudentiByPredmet(
 ticket: string,
 course: string,
 department: string,
): Promise<tGetStudentiByPredmet | null> {
 const url = new URL(`${process.env.STAG_SERVER}/services/rest2/student/getStudentiByPredmet`)
 url.searchParams.set('zkratka', course)
 url.searchParams.set('katedra', department)

 const headers = assembleHeaders(ticket)

 const res = await fetch(url.toString(), { method: 'GET', headers })
 if (!res.ok) return null
 if (res.status === 204) return null

 const data = (await res.json()) as tGetStudentiByPredmet
 for (const student of data.studentPredmetu) {
  student.encOsCislo = encodeId(student.osCislo)
 }
 return data
}

export async function getPredmetInfo(
 ticket: string,
 course: string,
 department: string,
): Promise<tGetPredmetInfo | null> {
 const url = new URL(`${process.env.STAG_SERVER}/services/rest2/predmety/getPredmetInfo`)
 url.searchParams.set('zkratka', course)
 url.searchParams.set('katedra', department)

 const headers = assembleHeaders(ticket)

 const res = await fetch(url.toString(), { method: 'GET', headers })
 if (!res.ok) return null
 if (res.status === 204) return null
 const data = await res.json()
 return data
}

export async function getStudentInfo(
 ticket: string,
 studentId: string,
): Promise<tGetStudentInfo | null> {
 const url = new URL(`${process.env.STAG_SERVER}/services/rest2/student/getStudentInfo`)
 url.searchParams.set('osCislo', studentId)

 const headers = assembleHeaders(ticket)

 const res = await fetch(url.toString(), { method: 'GET', headers })
 if (!res.ok) return null
 if (res.status === 204) return null

 const data = (await res.json()) as tGetStudentInfo
 if (!data) return null
 return data
}

export async function getRovrhByStudent(
 ticket: string,
 studentId: string,
): Promise<tGetRozvrhByStudent | null> {
 return getRozvrhByStudent(ticket, studentId)
}

export async function getRozvrhByStudent(
 ticket: string,
 studentId: string,
): Promise<tGetRozvrhByStudent | null> {
 const url = new URL(`${process.env.STAG_SERVER}/services/rest2/rozvrhy/getRozvrhByStudent`)
 url.searchParams.set('osCislo', studentId)

 const headers = assembleHeaders(ticket)

 const res = await fetch(url.toString(), { method: 'GET', headers })
 if (!res.ok) return null
 if (res.status === 204) return null

 const data = await res.json()
 if (!data) return null
 return data
}
